core/src/main/java/de/danoeh/antennapod/core/storage/DBTasks.java

```diff
@@ -373,118 +373,131 @@ private static FeedItem searchFeedItemByIdentifyingValue(Feed feed, String ident
-                            savedFeed.getItems().add(item);
-                        } else {
-                            savedFeed.getItems().add(idx, item);
-                        }
-
-                        // only mark the item new if it was published after or at the same time
-                        // as the most recent item
-                        // (if the most recent date is null then we can assume there are no items
-                        // and this is the first, hence 'new')
-                        if (priorMostRecentDate == null
-                                || priorMostRecentDate.before(item.getPubDate())
-                                || priorMostRecentDate.equals(item.getPubDate())) {
-                            Log.d(TAG, "Marking item published on " + item.getPubDate()
-                                    + " new, prior most recent date = " + priorMostRecentDate);
-                            item.setNew();
-                        }
-                    } else {
-                        oldItem.updateFromOther(item);
+            // identify items to be removed
+            if (removeUnlistedItems) {
+                Iterator<FeedItem> it = savedFeed.getItems().iterator();
+                while (it.hasNext()) {
+                    FeedItem feedItem = it.next();
+                    if (searchFeedItemByIdentifyingValue(newFeed, feedItem.getIdentifyingValue()) == null) {
+                        unlistedItems.add(feedItem);
+                        it.remove();
                     }
                 }
-                // update attributes
-                savedFeed.setLastUpdate(newFeed.getLastUpdate());
-                savedFeed.setType(newFeed.getType());
-                savedFeed.setLastUpdateFailed(false);
-
-                updatedFeedsList.add(savedFeed);
-                resultFeeds[feedIdx] = savedFeed;
             }
+
+            // update attributes
+            savedFeed.setLastUpdate(newFeed.getLastUpdate());
+            savedFeed.setType(newFeed.getType());
+            savedFeed.setLastUpdateFailed(false);
+
+            resultFeed = savedFeed;
         }
 
         adapter.close();
 
         try {
-            DBWriter.addNewFeed(context, newFeedsList.toArray(new Feed[0])).get();
-            DBWriter.setCompleteFeed(updatedFeedsList.toArray(new Feed[0])).get();
+            if (savedFeed == null) {
+                DBWriter.addNewFeed(context, newFeed).get();
+            } else {
+                DBWriter.setCompleteFeed(savedFeed).get();
+            }
+            if (removeUnlistedItems) {
+                DBWriter.deleteFeedItems(context, unlistedItems).get();
+            }
         } catch (InterruptedException | ExecutionException e) {
             e.printStackTrace();
         }
 
-        EventBus.getDefault().post(new FeedListUpdateEvent(updatedFeedsList));
+        if (savedFeed != null) {
+            EventBus.getDefault().post(new FeedListUpdateEvent(savedFeed));
```

Comment:
Could this lead to problems when adding a new feed? Looks like it might not appear in lists automatically, then.

Reply:
Oh, you are right. My refactoring changes the semantics in this case.

I updated the code to generate a `FeedListUpdateEvent` in any case.